include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -157,32 +161,291 @@ class NBL_API2 IShaderCompiler : public core::IReferenceCounted
+							if (definesIt->identifier != otherDefinesIt->identifier || definesIt->definition != otherDefinesIt->definition) return false;
+						}
+						return true;
+					}
+
+					// We don't need to hash the preprocessor on its own but in conjunction with the compiler options. We return data to perform the hash on from SCompilerOptions
+					inline std::vector<char> getHashable() const
+					{
+						std::vector<char> hashable;
+						hashable.insert(hashable.end(), sourceIdentifier.data()[0], sourceIdentifier.data()[sourceIdentifier.size()]);
+						core::vector<SMacroData> sortedExtraDefines;
+						sortedExtraDefines.assign(extraDefines.begin(), extraDefines.end());
+						// Sort them by identifier so the hash is not order-sensitive!
+						std::sort(sortedExtraDefines.begin(), sortedExtraDefines.end(), [](const SMacroData& lhs, const SMacroData& rhs) {return lhs.identifier < rhs.identifier; });
+						for (const auto& defines : sortedExtraDefines) {
+							hashable.insert(hashable.end(), defines.identifier.data()[0], defines.identifier.data()[defines.identifier.size()]);
+							hashable.insert(hashable.end(), defines.definition.data()[0], defines.definition.data()[defines.definition.size()]);
+						}
+
+						return hashable;
+					}
+
+					std::string sourceIdentifier;
+					std::vector<SMacroData> extraDefines;
+				};
+
+				struct SCompilerData {
+
+					inline bool operator==(const SCompilerData& other) const {
+						if (stage != other.stage || targetSpirvVersion != other.targetSpirvVersion || debugInfoFlags != other.debugInfoFlags || preprocessorData != other.preprocessorData) return false;
+						if (optimizerPasses.size() != other.optimizerPasses.size()) return false;
+						for (auto passesIt = optimizerPasses.begin(), otherPassesIt = other.optimizerPasses.begin(); passesIt != optimizerPasses.end(); passesIt++, otherPassesIt++) {
+							if (*passesIt != *otherPassesIt) return false;
+						}
+						return true;
+					}
+
+					inline std::vector<char> getHashable() const {
+						std::vector<char> hashable = preprocessorData.getHashable();
+						auto stageString = std::to_string(stage);
+						hashable.insert(hashable.end(), stageString.data()[0], stageString.data()[stageString.size()]);
+						auto versionString = std::to_string(static_cast<uint32_t>(targetSpirvVersion));
+						hashable.insert(hashable.end(), versionString.data()[0], versionString.data()[versionString.size()]);
+						auto debugString = std::to_string(static_cast<uint8_t>(debugInfoFlags.value));
+						hashable.insert(hashable.end(), debugString.data()[0], debugString.data()[debugString.size()]);
+						for (auto pass : optimizerPasses) {
+							auto passString = std::to_string(pass);
+							hashable.insert(hashable.end(), passString.data()[0], passString.data()[passString.size()]);
+						}
+						return hashable;
+					}
```

Comment:
don't even make this method, you need a precomputed hash, so:
- make the struct `final`
- make the members private
- provide a constructor, this constructor should do what `getHashable` does and store the xxHAsh256 result for later
- check the hash first in the `operator==` so its faster

P.S. Rememver that `stage` can be changed by the preprocessor (a `#pragma`) so may need to be updated after construction before the insertion.

Reply:
The PS poses an issue: if the stage is changed by the preprocessor, then the next time compilation is asked for with the same compiler options there won't be a match (because I'd need a preprocess to find out). I'm leaving it as it is right now, since if the stage is changed by a pragma then the contents of the file will still match

Reply:
you are right, make a comment about why its that way

(we're hashing serializing compiler inputs.... not results)